### ISSUE TITLE
[TF 2.0] GRU layer doesn't work when called from tf.function

### ISSUE THREAD

**Issue by David-Mao** [tensorflow/tensorflow#31998]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Darwin Kernel Version 18.6.0
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
N/A
- TensorFlow installed from (source or binary):
binary
- TensorFlow version (use command below):
2.0.0-dev20190826
- Python version:
Python 3.6.8 :: Anaconda, Inc.
- Bazel version (if compiling from source):
N/A
- GCC/Compiler version (if compiling from source):
N/A
- CUDA/cuDNN version:
N/A
- GPU model and memory:
N/A

**Describe the current behavior**

The code below raises error:

> Tried to convert 'tensor' to a tensor and failed. Error: None values not supported.

As noted in comments, the bug disappears if we don't use `tf.function`, or set `persistent=False` in gradient tape.

**Code to reproduce the issue**

```
import numpy as np
import tensorflow as tf

x = tf.cast(np.random.randn(1, 100), tf.float32)
y = tf.cast(np.random.randn(1, 100, 100), tf.float32)
z = tf.cast(np.random.randn(1, 100), tf.float32)

class Model(tf.keras.Model):
    def __init__(self):
        super(Model, self).__init__()
        self.layer = tf.keras.layers.GRU(100)

    @tf.function # remove this and it works fine
    def call(self, x, y):
        z = self.layer(y, initial_state=x)
        return z

model = Model()

with tf.GradientTape(persistent=True) as tape: # if persistent=False it works fine
    loss = tf.norm(model(x, y) - z)
grads = tape.gradient(loss, model.trainable_variables)
```

**Comment by oanush** [comment]
Issue replicating for TF version 2.0, please find the [Gist](https://colab.research.google.com/drive/1wvCV4_fGL9DHqyyVrLreTStmk0tqGcqb) of Colab.Thanks!

**Comment by allenlavoie** [comment]
Saurabh, looks WhileGrad / TensorList related?

**Comment by saxenasaurabh** [comment]
Yeah this seems to be failing when building the gradient of `TensorListPopBack`. It seems like with `persistent=True` this takes the path that outputs all necessary intermediates to support higher-order derivatives which triggers taking the derivative of `TensorListPopBack`. That path is not taken with `persistent=False` iiuc. I will look into a fix.

**Comment by amahendrakar** [comment]
@David-Mao,
I was able to reproduce the error with TF v2.0. However, the issue seems to be fixed with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/7033da4af6575764dcda2cf9b10ac6ed/2-1-template.ipynb). Please find the attached gist. Thanks!

**Comment by amahendrakar** [comment]
> @David-Mao,
> I was able to reproduce the error with TF v2.0. However, the issue seems to be fixed with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/7033da4af6575764dcda2cf9b10ac6ed/2-1-template.ipynb). Please find the attached gist. Thanks!

Any updates regarding this issue? Thanks!

**Comment by saxenasaurabh** [comment]
Sorry I don't understand. Should we close this since this seems to be fixed in 2.1? Are you asking if this will be fixed in 2.0? I don't think we would want to do that now since the fix is already available in a public release.

**Comment by StepNeverStop** [comment]
This problem still exists in TF2.1, the strangest thing is that the GRU is OK now but the LSTM is not. Does anybody have a solution?

```python
import numpy as np
import tensorflow as tf
from tensorflow.keras import Model as M
from tensorflow.keras import Input as I

x = tf.cast(np.random.randn(1, 100, 100), tf.float32)
y = tf.cast(np.random.randn(1, 100), tf.float32)
z = tf.cast(np.random.randn(1, 100), tf.float32)
u = tf.cast(np.random.randn(1, 100), tf.float32)

class Model(M):
    def __init__(self):
        super().__init__()
        self.layer = tf.keras.layers.GRU(100)
        self(I(shape=(None, 100)), I(shape=(100,)))

    @tf.function # remove this and it works fine
    def call(self, x, y):
        z = self.layer(x, initial_state=y)
        return z


model = Model()

with tf.GradientTape(persistent=True) as tape: # if persistent=False it works fine
    loss = tf.norm(model(x, y) - z)
grads = tape.gradient(loss, model.trainable_variables)

print("###############SUCCESS################")

class Model2(M):
    def __init__(self):
        super().__init__()
        self.layer = tf.keras.layers.LSTM(100)
        self(I(shape=(None, 100)), I(shape=(100,)), I(shape=(100,)))

    @tf.function     # remove this and it works fine
    def call(self, s, h, c):
        z = self.layer(s, initial_state=(h, c))
        return z


model2 = Model2()

with tf.GradientTape(persistent=True) as tape: # if persistent=False it works fine
    loss = tf.norm(model2(x, y, u) - z)
grads = tape.gradient(loss, model2.trainable_variables)


print("###############SUCCESS################")
```

**Comment by amahendrakar** [comment]
@StepNeverStop,
Could you please submit a new issue from [this link](https://github.com/tensorflow/tensorflow/issues/new/choose), so that we can track the issue there. Thanks!

**Comment by mdanatg** [comment]
BTW, I checked the code from @StepNeverStop with tf-nightly and the issue doesn't repro, so it might be that the fix landed in 2.3, instead of 2.2. It would be worth double checking.
